tests/generative/test_query_model.py

```diff
@@ -0,0 +1,357 @@
+#   in _all_ the abstract type strategies in order to allow us to keep the ordering the same as that of the definitions
+#   in `query_model`.
+# * The query model type hierarchy doesn't give us all the information we need, so the approach outlined above will
+#   generate invalid query model graphs (due either to the domains or the types being incorrect). We handle this by
+#   constructing the query model objects using a special-purpose strategy `qm_builds`, which can discard invalid
+#   instances. There is an unexplored optimization avenue here: make the strategies more specific to include type
+#   and/or domain, so that we don't discard so many examples.
+
+frame = st.deferred(
+    lambda: st.one_of(
+        one_row_per_patient_frame,
+        many_rows_per_patient_frame,
+    )
+)
+
+series = st.deferred(
+    lambda: st.one_of(
+        unknown_dimension_series,
+        definitely_one_row_patient_series,
+        definitely_many_rows_per_patient_series,
+    )
+)
+
+one_row_per_patient_frame = st.deferred(
+    lambda: st.one_of(
+        select_patient_table,
+        pick_one_row_per_patient,
+    )
+)
+
+many_rows_per_patient_frame = st.deferred(
+    lambda: st.one_of(
+        select_table,
+        sorted_frame,
+        filter,
+    )
+)
+
+one_row_per_patient_series = st.deferred(
+    lambda: st.one_of(
+        composite_filter(unknown_dimension_series, has_one_row_per_patient),
+        definitely_one_row_patient_series,
+    )
+)
+
+definitely_one_row_patient_series = st.deferred(
+    lambda: st.one_of(
+        value,
+        exists,
+        count_,
+        # TODO: not supported by in-memory engine
+        # min_,
+        # TODO: not supported by in-memory engine
+        # max_,
+        sum_,
+    )
+)
+
+many_rows_per_patient_series = st.deferred(
+    lambda: st.one_of(
+        composite_filter(unknown_dimension_series, has_many_rows_per_patient),
+        definitely_many_rows_per_patient_series,
+    )
+)
+
+# We don't currently have any examples of this type, but it's included here for consistency
+# and to make it clear where it fits in.
+definitely_many_rows_per_patient_series = st.deferred(lambda: st.one_of())
+
+unknown_dimension_series = st.deferred(
+    lambda: st.one_of(
+        select_column,
+        not_,
+        is_null,
+        negate,
+        eq,
+        ne,
+        lt,
+        le,
+        gt,
+        ge,
+        and_,
+        or_,
+        add,
+        subtract,
+    )
+)
+
+sorted_frame = st.deferred(lambda: st.one_of(sort))
+
+int_values = st.integers(min_value=0, max_value=10)
+bool_values = st.booleans()
+value = qm_builds(Value, st.one_of(int_values, bool_values))
+
+# To simplify data generation, all tables have the same schema.
+schema = TableSchema(i=int, b=bool)
+event_tables = {"e1": E1, "e2": E2}
+patient_tables = {"p1": P1, "p2": P2}
+
+select_table = qm_builds(
+    SelectTable, st.sampled_from(sorted(event_tables.keys())), st.just(schema)
```

Comment:
Without the explicit sort, I get this error from Hypohthesis:

> Cannot sample from dict_keys(['i1', 'i2', 'b1', 'b2']), not an ordered collection. Hypothesis goes to some length to ensure that the sampled_from strategy has stable results between runs. To replay a saved example, the sampled values must have the same iteration order on every run - ruling out sets, dicts, etc due to hash randomisation.

I guess that's a bug in Hypothesis, which hasn't been updated since dicts became stable.

But I've replaced `sorted()` with `list()` as a gesture of solidarity.

Reply:
Oh I see! Sorry to have created superfluous work